pkg/kubelet/pod_workers.go

```diff
@@ -221,61 +475,310 @@ func (p *podWorkers) UpdatePod(options *UpdatePodOptions) {
 			p.managePodLoop(podUpdates)
 		}()
 	}
+
+	// dispatch a request to the pod worker if none are running
 	if !p.isWorking[pod.UID] {
 		p.isWorking[pod.UID] = true
-		podUpdates <- *options
-	} else {
-		// if a request to kill a pod is pending, we do not let anything overwrite that request.
-		update, found := p.lastUndeliveredWorkUpdate[pod.UID]
-		if !found || update.UpdateType != kubetypes.SyncPodKill {
-			p.lastUndeliveredWorkUpdate[pod.UID] = *options
+		podUpdates <- options
+		return
+	}
+
+	// never lose an OnCompleteFunc or a PodStatusFunc
+	if undelivered, ok := p.lastUndeliveredWorkUpdate[pod.UID]; ok {
+		if undelivered.OnCompleteFunc != nil {
```

Comment:
If there are two kills (delete at 30s then evict at 1s), and the second one shortens, we want to terminate the first on complete.  Or at least, that's why I was preserving it.  I think OnCompleteFunc is a misfeature and we should instead accept a channel that we guarantee close at a certain lifecycle point - that would fix killPodNow and also be safer (the use of OnCompleteFunc in UpdatePod is really wrong anyway and sync metrics need to be calculated by the worker looking at podSyncStatus)

Reply:
I went back and looked at killPodNow.  We really didn't need OnCompleteFunc and it is complex.  I instead put something more specific in (which is an optional channel on KillPodOptions that is closed when we finally transition from terminating -> terminated).  Also moved sync worker inline.  PodStatusFunc still has to be accumulated, but now only KillPodOptions are accumulated.  PodStatusFunc is accumulative because you could preempt then evict (which I think is correct), but could be more complex.